### PR TITLE
core: fix potential misorder of latch waiters

### PR DESCRIPTION
**Pull request by Gumix**
#### core: fix potential misorder of latch waiters

Currently the latch doesn't guarantee the order in which it is acquired
by fibers, which requested it. E.g. it is possible to wake up spuriously
a fiber which is yielding in the latch_lock, it will be removed from
l->queue by fiber_make_ready, then it will be inserted to l->queue
again, but for this time, to the head of the list instead of its
original place in the queue.

Fix this by using latch_waiter structure, which is linked into l->queue.

Part of #7166

### CODE REVIEW

File: src/box/txn_limbo.h [hunk 1]
```diff
@@ -266,6 +276,59 @@ txn_limbo_replica_term(const struct txn_limbo *limbo, uint32_t replica_id)
...
+	}
+
+	waiter->owner = fiber();
+	rlist_add_tail_entry(queue, waiter, queue);
+	do {
+		fiber_yield();
+	} while (limbo->promote_lock.owner != NULL);
+
+	limbo->promote_lock.owner = fiber();
+	rlist_del_entry(waiter, queue);
```
Comment (locker) [inline src/box/txn_limbo.h:311]: Looks like this code could be reused somewhere else. Maybe, add a generic data structure for this ordered latch?

Comment (sergepetrenko) [reply src/box/txn_limbo.h:311]: > Looks like this code could be reused somewhere else. Maybe, add a generic data structure for this ordered latch?

+1

**Review by sergepetrenko — CHANGES_REQUESTED** [review]
Thanks for the patch!
Please find a couple of comments below.

File: src/box/applier.cc [hunk 2]
```diff
@@ -1017,7 +1017,8 @@ apply_synchro_req(uint32_t replica_id, struct xrow_header *row, struct synchro_r
 	 * before trying to commit. But that requires extra steps from the
 	 * transactions side, including the async ones.
 	 */
-	txn_limbo_begin(&txn_limbo);
+	struct txn_limbo_promote_lock lock;
```
[the remark this answers is not in the document]
Comment (Gumix) [reply src/box/applier.cc:1020]: This struct is declared outside of `txn_limbo_begin`, because it is added to the list and should be alive till `txn_limbo_{commit|rollback}`.

[the remark this answers is not in the document]
Comment (sergepetrenko) [reply src/box/applier.cc:1020]: Am I missing something?
You remove the waiter from the list right after acquiring the lock in `txn_limbo_promote_lock()`.

Why do you need the struct to be alive till `txn_limbo_commit/rollback` then?

File: src/box/txn_limbo.h (hunk 1 above)
Comment (Gumix) [reply src/box/txn_limbo.h]: Done.

File: src/lib/core/ordered_latch.h [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+#include "small/rlist.h"
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+/**
+ * Latch of cooperative multitasking environment, which preserves strict order
+ * of fibers waiting for the latch.
+ */
```
Comment (locker) [inline src/lib/core/ordered_latch.h:19]: Why should one prefer to use `latch` instead of `ordered_latch` if they don't need ordering guarantees?

I thought that `ordered_latch` would have a different, less convenient API, but it looks like it has exactly the same API as `latch` and doesn't have any performance overhead in comparison to simple `latch`. Maybe, just rework the existing `latch` as this `ordered_latch`?

File: src/lib/core/ordered_latch.h [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+static inline void
+ordered_latch_lock(struct ordered_latch *l)
+{
+	assert(l->owner != fiber());
+	if (l->owner == NULL && rlist_empty(&l->queue)) {
+		l->owner = fiber();
+		return;
+	}
+
+	struct ordered_latch waiter;
```
Comment (locker) [inline src/lib/core/ordered_latch.h:71]: It's confusing to use the same struct for a latch and a waiter IMO. Also, `waiter->owner` sounds misleading - it should probably be `waiter->self`. Better introduce a separate structure.

File: src/lib/core/ordered_latch.h [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	waiter.owner = fiber();
+	rlist_add_tail_entry(&l->queue, &waiter, queue);
+	do {
+		fiber_yield();
```
Comment (sergepetrenko) [inline src/lib/core/ordered_latch.h:75]: Should it be possible to cancel a fiber trying to acquire the latch?
Currently it will simply ignore the cancellation and wait forever potentially.

File: src/lib/core/ordered_latch.h [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	} while (l->owner != NULL);
```
Comment (sergepetrenko) [inline src/lib/core/ordered_latch.h:76]: You made it so that waking a fiber up doesn't put it in front of the queue, but waking up some fiber from the end of the queue will still make it steal the latch. I think there should be some check for whether this waiter is the first in queue or not.
The test fails after the following change:
```diff
diff --git a/test/unit/ordered_latch.c b/test/unit/ordered_latch.c
index 79dd6eea0..a455a99a7 100644
--- a/test/unit/ordered_latch.c
+++ b/test/unit/ordered_latch.c
@@ -36,13 +36,13 @@ main_f(va_list ap)
                fiber_set_joinable(fibers[i], true);
                fiber_start(fibers[i], &i, &check, &latch);
        }
-
-       ordered_latch_unlock(&latch);
        /*
         * Try to break the order of waiters on the latch.
         */
        fiber_wakeup(fibers[1]);
 
+       ordered_latch_unlock(&latch);
+
        for (size_t i = 0; i < num_fibers; i++)
                fiber_join(fibers[i]);
        ordered_latch_destroy(&latch);
```
```
[001] unit/ordered_latch.test                                         [ fail ]
[001] 
[001] Test failed! Result content mismatch:
[001] --- unit/ordered_latch.result	Fri Jul  8 09:49:30 2022
[001] +++ /tmp/t/rejects/unit/ordered_latch.reject	Fri Jul  8 10:02:15 2022
[001] @@ -1,7 +1,12 @@
[001] +#   Failed test 'check order'
[001] +#   in /Users/s.petrenko/Source/sptnt/test/unit/ordered_latch.c at line 17
[001] +#   Failed test 'check order'
[001] +#   in /Users/s.petrenko/Source/sptnt/test/unit/ordered_latch.c at line 17
[001] +# Looks like you failed 2 test of 3 run.
[001]  1..3
[001] +fiber 1 acquired the latch
[001] +not ok 1 - check order
[001]  fiber 0 acquired the latch
[001] -ok 1 - check order
[001] -fiber 1 acquired the latch
[001] -ok 2 - check order
[001] +not ok 2 - check order
[001]  fiber 2 acquired the latch
[001]  ok 3 - check order
[Main process] Got failed test; gently terminate all workers...
[001] Worker "001_unit" got failed test; stopping the server...```

File: src/lib/core/ordered_latch.h (hunk 5 above)
Comment (Gumix) [reply src/lib/core/ordered_latch.h:75]: I am not sure… @locker?

File: src/lib/core/ordered_latch.h (hunk 6 above)
Comment (Gumix) [reply src/lib/core/ordered_latch.h:76]: Thanks for the catch.

File: src/lib/core/ordered_latch.h (hunk 5 above)
Comment (locker) [reply src/lib/core/ordered_latch.h:75]: AFAICS it's impossible to cancel a fiber waiting on a latch in the current implementation, because we make it non-cancellable before yield:
https://github.com/tarantool/tarantool/blob/9c87898ea0fa9d6ffd721187b9c17ddf4d9bc0bc/src/lib/core/latch.h#L131

I don't see why we should allow that in the new implementation.

File: src/lib/core/ordered_latch.h (hunk 3 above)
Comment (Gumix) [reply src/lib/core/ordered_latch.h]: Done

File: src/lib/core/ordered_latch.h (hunk 4 above)
Comment (Gumix) [reply src/lib/core/ordered_latch.h]: Done

**Review by locker — APPROVED** [review]

File: src/lib/core/latch.h [hunk 7]
```diff
@@ -127,24 +144,24 @@ latch_lock_timeout(struct latch *l, ev_tstamp timeout)
 	if (timeout <= 0)
 		return 1;
 
-	rlist_add_tail_entry(&l->queue, fiber(), state);
-	bool was_cancellable = fiber_set_cancellable(false);
```
Comment (sergepetrenko) [inline src/lib/core/latch.h:131]: Let's split removing `fiber_set_cancellable` and other changes in two commits.
Removing `fiber_set_cancellable` seems unrelated to fixing latch order.

**Review by sergepetrenko — CHANGES_REQUESTED** [review]
Thanks for the fixes!
Please, address one final comment.

**Review by sergepetrenko — APPROVED** [review]
Thanks! LGTM.

File: test/unit/latch.c [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+static int
+order_f(va_list ap)
+{
+	size_t fid = *va_arg(ap, size_t *);
+	size_t *check = va_arg(ap, size_t *);
+	struct latch *latch = va_arg(ap, struct latch *);
+	latch_lock(latch);
+
+	printf("fiber %zu acquired the latch\n", fid);
```
Comment (locker) [inline test/unit/latch.c:16]: Debug print, should be removed? It messes with the test output.

Comment (Gumix) [reply test/unit/latch.c]: Ok, removed.
